cpp/src_prims/distance/pairwise_distance_base.cuh

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include <raft/linalg/contractions.cuh>
+#include <raft/linalg/norm.cuh>
+
+namespace MLCommon {
+namespace Distance {
+
+/**
+ * @brief Device class for L2 expanded, unexpanded distance metrics.
+ * @tparam DataT input data-type (for A and B matrices)
+ * @tparam AccT accumulation data-type
+ * @tparam OutT output data-type (for C and D matrices)
+ * @tparam IdxT index data-type
+ * @tparam bool Sqrt if the square root is computed or not
+ * @tparam Policy struct which tunes the Contraction kernel
+ * @tparam CoreLambda, tells how to accumulate an x and y into 
+    acc. its signature:
+    template <typename AccT, typename DataT> void core_lambda(AccT& acc,
+      const DataT& x, const DataT& y)
+ * @tparam EpilogueLambda applies an elementwise function to compute final 
+    values. Its signature is:
+    template <typename AccT, typename DataT> void epilogue_lambda
+    (AccT acc[][], DataT* sxNorm, DataT* sxNorm);
+ * @tparam FinalLambda the final lambda called on final distance value
+ * @param[in] x input matrix
+ * @param[in] y input matrix
+ * @param[in] m number of rows of A and C/D
+ * @param[in] n number of columns of B and C/D
+ * @param[in] k number of cols of A and rows of B
+ * @param[in] xn row norms of input matrix A. Required for expanded L2
+ * @param[in] yn row norms of input matrix B. Required for expanded L2
+ * @param[output] pD output matrix
+ * @param[in] smem shared mem buffer for intermediate storage of A, B, xn & yn.
+ * @param core_op the core accumulation operation lambda
+ * @param epilog_op the epilog operation lambda
+ * @param fin_op the final gemm epilogue lambda
```

Comment:
Just curious, why do we need a `fin_op`? Why can't it be done inside `epilog_op` itself?

Reply:
fin_op is for any additional operations user wants to perform on final acc values, so IMO this should be kept different from epilog_op.

Reply:
Fair enough. Makes sense to keep it separate.